crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -0,0 +1,600 @@
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    path::{Path, PathBuf},
+};
+
+use async_stream::try_stream;
+pub use futures::{pin_mut, Stream, StreamExt};
+use inotify::{EventMask, Inotify, WatchMask};
+use maplit::{btreeset, hashmap};
+use strum_macros::Display;
+use try_traits::default::TryDefault;
+
+const DIRECTORY_IS_WATCHED: &str = "C8Y_DIR_WATCH";
+
+#[derive(Debug, Display, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub enum FileEvent {
+    Modified,
+    Deleted,
+    Created,
+}
+
+impl TryFrom<FileEvent> for WatchMask {
+    type Error = NotifyStreamError;
+
+    fn try_from(value: FileEvent) -> Result<Self, Self::Error> {
+        match value {
+            FileEvent::Modified => Ok(WatchMask::MODIFY),
+            FileEvent::Deleted => Ok(WatchMask::DELETE),
+            FileEvent::Created => Ok(WatchMask::CREATE),
+        }
+    }
+}
+
+impl TryFrom<EventMask> for FileEvent {
+    type Error = NotifyStreamError;
+
+    fn try_from(value: EventMask) -> Result<Self, Self::Error> {
+        match value {
+            EventMask::MODIFY => Ok(FileEvent::Modified),
+            EventMask::DELETE => Ok(FileEvent::Deleted),
+            EventMask::CREATE => Ok(FileEvent::Created),
+            _ => Err(NotifyStreamError::UnsupportedEventMask { mask: value }),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum NotifyStreamError {
+    #[error(transparent)]
+    FromIOError(#[from] std::io::Error),
+
+    #[error("Error starting fs notification service.")]
```

Comment:
I don't think we have a concrete rule if the sentence should end with `.` or not. However, I see most of our errors end without `.`, so I want to keep some consistency here.

```suggestion
    #[error("Error starting fs notification service")]
```

Reply:
Nice catch, missed that! :wink: 

From my experience, error messages do never end with a period, so :+1: for catching this!